# Keep symlinks as symlinks when FTP RNTO / SFTP RENAME falls back to copying
# across filesystem mounts (the fallback followed source and destination links).
# Upstream Issue #2256: https://github.com/proftpd/proftpd/issues/2256
# Commits (master only, not in 1.3.9e): bdb4901, 8255a02, 7610f88, 0fa0f97
diff --git a/contrib/mod_copy.c b/contrib/mod_copy.c
index a87a436..a807c2f 100644
--- a/contrib/mod_copy.c
+++ b/contrib/mod_copy.c
@@ -157,37 +157,6 @@ static int create_path(pool *p, const char *path) {
   return 0;
 }
 
-static int copy_symlink(pool *p, const char *src_path, const char *dst_path,
-    int flags) {
-  char *link_path = pcalloc(p, PR_TUNABLE_BUFFER_SIZE);
-  int len;
-
-  len = pr_fsio_readlink(src_path, link_path, PR_TUNABLE_BUFFER_SIZE-1);
-  if (len < 0) {
-    int xerrno = errno;
-
-    pr_log_pri(PR_LOG_WARNING, MOD_COPY_VERSION ": error reading link '%s': %s",
-      src_path, strerror(xerrno));
-
-    errno = xerrno;
-    return -1;
-  }
-  link_path[len] = '\0';
-
-  if (pr_fsio_symlink(link_path, dst_path) < 0) {
-    int xerrno = errno;
-
-    pr_log_pri(PR_LOG_WARNING, MOD_COPY_VERSION
-      ": error symlinking '%s' to '%s': %s", link_path, dst_path,
-      strerror(xerrno));
-
-    errno = xerrno;
-    return -1;
-  }
-
-  return 0;
-}
-
 static int copy_dir(pool *p, const char *src_dir, const char *dst_dir,
     int flags) {
   DIR *dh = NULL;
@@ -317,7 +286,7 @@ static int copy_dir(pool *p, const char *src_dir, const char *dst_dir,
 
     /* Is this path a symlink? */
     } else if (S_ISLNK(st.st_mode)) {
-      if (copy_symlink(iter_pool, src_path, dst_path, flags) < 0) {
+      if (pr_fs_copy_symlink(src_path, dst_path, flags) < 0) {
         res = -1;
         break;
       }
@@ -469,9 +438,13 @@ static int copy_paths(pool *p, const char *from, const char *to) {
         errno = EACCES;
         return -1;
       }
+
+    } else {
+      pr_trace_msg(trace_channel, 19,
+        "lstat(2) error for destination path '%s': %s", to, strerror(errno));
     }
 
-    res = copy_symlink(p, from, to, flags);
+    res = pr_fs_copy_file2(from, to, flags, NULL);
     if (res < 0) {
       int xerrno = errno;
 
diff --git a/contrib/mod_sftp/fxp.c b/contrib/mod_sftp/fxp.c
index 6de598d..a15361b 100644
--- a/contrib/mod_sftp/fxp.c
+++ b/contrib/mod_sftp/fxp.c
@@ -5377,6 +5377,118 @@ static int fxp_handle_ext_limits(struct fxp_packet *fxp) {
   return fxp_packet_write(resp);
 }
 
+static int rename_across_mounts(pool *p, const char *src_path,
+    const char *dst_path) {
+  int res, xerrno;
+  struct stat src_st, dst_st;
+
+  pr_fs_clear_cache2(src_path);
+  res = pr_fsio_lstat(src_path, &src_st);
+  if (res < 0) {
+    xerrno = errno;
+
+    pr_trace_msg(trace_channel, 3, "rename: lstat(2) error on '%s': %s",
+      src_path, strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  if (S_ISDIR(src_st.st_mode)) {
+    xerrno = EISDIR;
+
+    pr_trace_msg(trace_channel, 3,
+      "rename: cannot rename '%s' across mounts: %s", src_path,
+      strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  if (!S_ISREG(src_st.st_mode) &&
+      !S_ISLNK(src_st.st_mode)) {
+    xerrno = EPERM;
+
+    pr_trace_msg(trace_channel, 3,
+      "rename: cannot rename unsupported file type '%s' across mounts: %s",
+      src_path, strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  pr_fs_clear_cache2(dst_path);
+  res = pr_fsio_lstat(dst_path, &dst_st);
+  if (res < 0 &&
+      errno != ENOENT) {
+    xerrno = errno;
+
+    pr_trace_msg(trace_channel, 3, "rename: lstat(2) error on '%s': %s",
+      dst_path, strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  if (res == 0) {
+    if (S_ISDIR(dst_st.st_mode)) {
+      xerrno = EPERM;
+
+      pr_trace_msg(trace_channel, 3,
+        "rename: cannot rename '%s' across mounts into directory '%s': %s",
+        src_path, dst_path, strerror(xerrno));
+
+      errno = xerrno;
+      return -1;
+    }
+
+    res = pr_fsio_unlink(dst_path);
+    if (res < 0) {
+      xerrno = errno;
+
+      pr_trace_msg(trace_channel, 3, "rename: unlink(2) error on '%s': %s",
+        dst_path, strerror(xerrno));
+
+      errno = xerrno;
+      return -1;
+    }
+  }
+
+  /* At this point, we know that the source path exists, and is a regular file
+   * or a symlink.  And we know that the destination path does not exist.
+   */
+
+  if (S_ISREG(src_st.st_mode)) {
+    res = pr_fs_copy_file(src_path, dst_path);
+    xerrno = errno;
+
+  } else if (S_ISLNK(src_st.st_mode)) {
+    res = pr_fs_copy_symlink(src_path, dst_path, 0);
+    xerrno = errno;
+  }
+
+  if (res < 0) {
+    (void) pr_trace_msg("fileperms", 1, "RENAME, user '%s' (UID %s, "
+      "GID %s): error copying '%s' to '%s': %s", session.user,
+      pr_uid2str(p, session.uid), pr_gid2str(p, session.gid), src_path,
+      dst_path, strerror(xerrno));
+
+    (void) pr_log_writefile(sftp_logfd, MOD_SFTP_VERSION,
+      "error copying '%s' to '%s': %s", src_path, dst_path, strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  /* Once copied, remove the original path. */
+  if (pr_fsio_unlink(src_path) < 0) {
+    (void) pr_log_writefile(sftp_logfd, MOD_SFTP_VERSION,
+      "error deleting '%s': %s", src_path, strerror(errno));
+  }
+
+  return 0;
+}
+
 static int fxp_handle_ext_posix_rename(struct fxp_packet *fxp, char *src,
     char *dst) {
   unsigned char *buf, *ptr;
@@ -5638,7 +5750,7 @@ static int fxp_handle_ext_posix_rename(struct fxp_packet *fxp, char *src,
        */
       errno = 0;
 
-      res = pr_fs_copy_file2(src, dst, 0, NULL);
+      res = rename_across_mounts(fxp->pool, src, dst);
       if (res < 0) {
         xerrno = errno;
 
@@ -5652,15 +5764,6 @@ static int fxp_handle_ext_posix_rename(struct fxp_packet *fxp, char *src,
           "error copying '%s' to '%s': %s", src, dst, strerror(xerrno));
 
         errno = xerrno;
-
-      } else {
-        /* Once copied, remove the original path. */
-        if (pr_fsio_unlink(src) < 0) {
-          (void) pr_log_writefile(sftp_logfd, MOD_SFTP_VERSION,
-            "error deleting '%s': %s", src, strerror(errno));
-        }
-
-        xerrno = errno = 0;
       }
     }
 
@@ -5676,7 +5779,7 @@ static int fxp_handle_ext_posix_rename(struct fxp_packet *fxp, char *src,
     xerrno != EOF ? strerror(xerrno) : "End of file", xerrno);
 
   /* Clear out any transfer-specific data. */
-  if (session.xfer.p) {
+  if (session.xfer.p != NULL) {
     destroy_pool(session.xfer.p);
   }
   memset(&session.xfer, 0, sizeof(session.xfer));
@@ -12395,33 +12498,25 @@ static int fxp_handle_rename(struct fxp_packet *fxp) {
       errno = xerrno;
 
     } else {
+      errno = 0;
+
       /* In this case, we should manually copy the file from the source
        * path to the destination path.
        */
-      errno = 0;
-      if (pr_fs_copy_file2(old_path, new_path, 0, NULL) < 0) {
+      if (rename_across_mounts(fxp->pool, old_path, new_path) < 0) {
         xerrno = errno;
 
         (void) pr_trace_msg("fileperms", 1, "RENAME, user '%s' (UID %s, "
-          "GID %s): error copying '%s' to '%s': %s", session.user,
+          "GID %s): error renaming '%s' to '%s': %s", session.user,
           pr_uid2str(fxp->pool, session.uid),
           pr_gid2str(fxp->pool, session.gid), old_path, new_path,
           strerror(xerrno));
 
         (void) pr_log_writefile(sftp_logfd, MOD_SFTP_VERSION,
-          "error copying '%s' to '%s': %s", old_path, new_path,
-          strerror(xerrno));
+          "RENAME of '%s' to '%s' across mounts failed: %s", old_path,
+          new_path, strerror(xerrno));
 
         errno = xerrno;
-
-      } else {
-        /* Once copied, remove the original path. */
-        if (pr_fsio_unlink(old_path) < 0) {
-          (void) pr_log_writefile(sftp_logfd, MOD_SFTP_VERSION,
-            "error deleting '%s': %s", old_path, strerror(errno));
-        }
-
-        xerrno = errno = 0;
       }
     }
 
@@ -12437,7 +12532,7 @@ static int fxp_handle_rename(struct fxp_packet *fxp) {
     xerrno != EOF ? strerror(xerrno) : "End of file", xerrno);
 
   /* Clear out any transfer-specific data. */
-  if (session.xfer.p) {
+  if (session.xfer.p != NULL) {
     destroy_pool(session.xfer.p);
   }
   memset(&session.xfer, 0, sizeof(session.xfer));
diff --git a/include/fsio.h b/include/fsio.h
index 5ac0018..b9b184a 100644
--- a/include/fsio.h
+++ b/include/fsio.h
@@ -414,6 +414,11 @@ int pr_fs_copy_file2(const char *src, const char *dst, int flags,
   void (*progress_cb)(int));
 #define PR_FSIO_COPY_FILE_FL_NO_DELETE_ON_FAILURE	0x0001
 
+/* Copy a symlink from the given source path to the destination path, creating
+ * a symlink pointing to the same file at the destination.
+ */
+int pr_fs_copy_symlink(const char *src, const char *dst, int flags);
+
 int pr_fs_setcwd(const char *);
 const char *pr_fs_getcwd(void);
 const char *pr_fs_getvwd(void);
diff --git a/modules/mod_core.c b/modules/mod_core.c
index e2b38ae..18766f3 100644
--- a/modules/mod_core.c
+++ b/modules/mod_core.c
@@ -6134,8 +6134,127 @@ MODRET core_dele(cmd_rec *cmd) {
   return PR_HANDLED(cmd);
 }
 
+static int handle_cross_mount_rename(cmd_rec *cmd, const char *src_path,
+    const char *dst_path, struct stat *dst_st) {
+  int res, xerrno;
+  struct stat src_st;
+
+  pr_fs_clear_cache2(src_path);
+  res = pr_fsio_lstat(src_path, &src_st);
+  if (res < 0) {
+    xerrno = errno;
+
+    pr_log_debug(DEBUG4, "Unable to stat '%s' for cross filesystem mount "
+      "point rename: %s", src_path, strerror(xerrno));
+
+    pr_response_add_err(R_550, _("%s: %s"), cmd->arg, strerror(xerrno));
+    pr_cmd_set_errno(cmd, xerrno);
+    errno = xerrno;
+
+    return -1;
+  }
+
+  if (S_ISDIR(src_st.st_mode)) {
+    xerrno = EPERM;
+
+    pr_log_debug(DEBUG4, "Cannot rename directory '%s' across a filesystem "
+      "mount", src_path);
+
+    pr_response_add_err(R_550, _("%s: %s"), cmd->arg, strerror(xerrno));
+    pr_cmd_set_errno(cmd, xerrno);
+    errno = xerrno;
+
+    return -1;
+  }
+
+  if (!S_ISREG(src_st.st_mode) &&
+      !S_ISLNK(src_st.st_mode)) {
+    xerrno = EPERM;
+
+    pr_log_debug(DEBUG4, "Cannot rename unsupported file type '%s' across "
+      "a filesystem mount", src_path);
+
+    pr_response_add_err(R_550, _("%s: %s"), cmd->arg, strerror(xerrno));
+    pr_cmd_set_errno(cmd, xerrno);
+    errno = xerrno;
+
+    return -1;
+  }
+
+  /* If the destination exists, unlink it first.  Unless it is a directory,
+   * in which case we error out now; we do not copy into directories across
+   * mounts at the moment.
+   */
+  if (dst_st != NULL) {
+    if (S_ISDIR(dst_st->st_mode)) {
+      xerrno = EPERM;
+
+      pr_log_debug(DEBUG4, "Cannot rename path '%s' across a filesystem mount "
+        "point into directory '%s'", src_path, dst_path);
+
+      pr_response_add_err(R_550, _("%s: %s"), cmd->arg, strerror(xerrno));
+      pr_cmd_set_errno(cmd, xerrno);
+      errno = xerrno;
+
+      return -1;
+    }
+
+    res = pr_fsio_unlink(dst_path);
+    if (res < 0) {
+      xerrno = errno;
+
+      pr_log_debug(DEBUG4, "Unable to delete '%s' for cross filesystem mount "
+        "point rename: %s", dst_path, strerror(xerrno));
+
+      pr_response_add_err(R_550, _("%s: %s"), cmd->arg, strerror(xerrno));
+      pr_cmd_set_errno(cmd, xerrno);
+      errno = xerrno;
+
+      return -1;
+    }
+
+    dst_st = NULL;
+  }
+
+  /* At this point, we know that the source path exists, and is a regular file
+   * or symlink.  And we know that the destination path does not exist.
+   */
+
+  if (S_ISREG(src_st.st_mode)) {
+    res = pr_fs_copy_file(src_path, dst_path);
+    xerrno = errno;
+
+  } else if (S_ISLNK(src_st.st_mode)) {
+    res = pr_fs_copy_symlink(src_path, dst_path, 0);
+    xerrno = errno;
+  }
+
+  if (res < 0) {
+    (void) pr_trace_msg("fileperms", 1, "%s, user '%s' (UID %s, GID %s): "
+      "error copying '%s' to '%s': %s", (char *) cmd->argv[0], session.user,
+      pr_uid2str(cmd->tmp_pool, session.uid),
+      pr_gid2str(cmd->tmp_pool, session.gid), src_path, dst_path,
+      strerror(xerrno));
+
+    pr_response_add_err(R_550, _("Rename %s: %s"), cmd->arg,
+      strerror(xerrno));
+    pr_cmd_set_errno(cmd, xerrno);
+    errno = xerrno;
+
+    return -1;
+  }
+
+  /* Once copied, unlink the original file. */
+  res = pr_fsio_unlink(src_path);
+  if (res < 0) {
+    pr_log_debug(DEBUG0, "error deleting '%s': %s", src_path, strerror(errno));
+  }
+
+  return 0;
+}
+
 MODRET core_rnto(cmd_rec *cmd) {
-  int res;
+  int res, xerrno, dst_exists = FALSE;
   char *decoded_path, *path;
   unsigned char *allow_overwrite = NULL;
   struct stat st;
@@ -6159,7 +6278,7 @@ MODRET core_rnto(cmd_rec *cmd) {
   decoded_path = pr_fs_decode_path2(cmd->tmp_pool, cmd->arg,
     FSIO_DECODE_FL_TELL_ERRORS);
   if (decoded_path == NULL) {
-    int xerrno = errno;
+    xerrno = errno;
 
     pr_log_debug(DEBUG8, "'%s' failed to decode properly: %s", cmd->arg,
       strerror(xerrno));
@@ -6199,23 +6318,28 @@ MODRET core_rnto(cmd_rec *cmd) {
 
   path = dir_canonical_path(cmd->tmp_pool, path);
 
-  allow_overwrite = get_param_ptr(CURRENT_CONF, "AllowOverwrite", FALSE);
-
   /* Deny the rename if AllowOverwrites are not allowed, and the destination
    * rename file already exists.
    */
+  allow_overwrite = get_param_ptr(CURRENT_CONF, "AllowOverwrite", FALSE);
+
   pr_fs_clear_cache2(path);
-  if ((!allow_overwrite || *allow_overwrite == FALSE) &&
-      pr_fsio_stat(path, &st) == 0) {
-    pr_log_debug(DEBUG6, "AllowOverwrite denied permission for %s", path);
-    pr_response_add_err(R_550, _("%s: Rename permission denied"), cmd->arg);
+  res = pr_fsio_lstat(path, &st);
+  if (res == 0) {
+    dst_exists = TRUE;
 
-    pr_cmd_set_errno(cmd, EACCES);
-    errno = EACCES;
-    return PR_ERROR(cmd);
+    if (allow_overwrite == NULL ||
+        *allow_overwrite == FALSE) {
+      pr_log_debug(DEBUG6, "AllowOverwrite denied permission for %s", path);
+      pr_response_add_err(R_550, _("%s: Rename permission denied"), cmd->arg);
+
+      pr_cmd_set_errno(cmd, EACCES);
+      errno = EACCES;
+      return PR_ERROR(cmd);
+    }
   }
 
-  if (!path ||
+  if (path == NULL ||
       !dir_check_canon(cmd->tmp_pool, cmd, cmd->group, path, NULL)) {
     pr_log_debug(DEBUG8, "%s command denied by <Limit> config",
       (char *) cmd->argv[0]);
@@ -6228,7 +6352,7 @@ MODRET core_rnto(cmd_rec *cmd) {
 
   res = pr_fsio_rename_with_error(cmd->pool, session.xfer.path, path, &err);
   if (res < 0) {
-    int xerrno = errno;
+    xerrno = errno;
 
     pr_error_set_where(err, &core_module, __FILE__, __LINE__ - 4);
     pr_error_set_why(err, pstrcat(cmd->pool, "rename '", session.xfer.path,
@@ -6295,42 +6419,10 @@ MODRET core_rnto(cmd_rec *cmd) {
     /* In this case, we'll need to manually copy the file from the source
      * to the destination paths.
      */
-    if (pr_fs_copy_file(session.xfer.path, path) < 0) {
-      xerrno = errno;
-
-      (void) pr_trace_msg("fileperms", 1, "%s, user '%s' (UID %s, GID %s): "
-        "error copying '%s' to '%s': %s", (char *) cmd->argv[0], session.user,
-        pr_uid2str(cmd->tmp_pool, session.uid),
-        pr_gid2str(cmd->tmp_pool, session.gid), session.xfer.path, path,
-        strerror(xerrno));
-
-      pr_response_add_err(R_550, _("Rename %s: %s"), cmd->arg,
-        strerror(xerrno));
-
-      pr_cmd_set_errno(cmd, xerrno);
-      errno = xerrno;
+    if (handle_cross_mount_rename(cmd, session.xfer.path, path,
+        dst_exists == TRUE ? &st : NULL) < 0) {
       return PR_ERROR(cmd);
     }
-
-    /* Once copied, unlink the original file. */
-    res = pr_fsio_unlink_with_error(cmd->pool, session.xfer.path, &err);
-    if (res < 0) {
-      xerrno = errno;
-
-      pr_error_set_where(err, &core_module, __FILE__, __LINE__ - 4);
-      pr_error_set_why(err, pstrcat(cmd->pool, "delete file '",
-        session.xfer.path, "'", NULL));
-
-      if (err != NULL) {
-        pr_log_debug(DEBUG0, "%s", pr_error_strerror(err, 0));
-        pr_error_destroy(err);
-        err = NULL;
-
-      } else {
-        pr_log_debug(DEBUG0, "error deleting '%s': %s", session.xfer.path,
-          strerror(xerrno));
-      }
-    }
   }
 
   /* Change the xfer path to the name of the destination file, for logging. */
diff --git a/src/fsio.c b/src/fsio.c
index 43c47b7..53789bd 100644
--- a/src/fsio.c
+++ b/src/fsio.c
@@ -2235,6 +2235,45 @@ int pr_fs_copy_file(const char *src, const char *dst) {
   return pr_fs_copy_file2(src, dst, 0, NULL);
 }
 
+int pr_fs_copy_symlink(const char *src, const char *dst, int flags) {
+  char link_path[PR_TUNABLE_BUFFER_SIZE];
+  int len;
+
+  /* Currently unused. */
+  (void) flags;
+
+  if (src == NULL ||
+      dst == NULL) {
+    errno = EINVAL;
+    return -1;
+  }
+
+  memset(link_path, '\0', sizeof(link_path));
+  len = pr_fsio_readlink(src, link_path, sizeof(link_path)-1);
+  if (len < 0) {
+    int xerrno = errno;
+
+    pr_log_pri(PR_LOG_WARNING, "error reading symlink '%s': %s", src,
+      strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+  link_path[len] = '\0';
+
+  if (pr_fsio_symlink(link_path, dst) < 0) {
+    int xerrno = errno;
+
+    pr_log_pri(PR_LOG_WARNING, "error symlinking '%s' to '%s': %s", link_path,
+      dst, strerror(xerrno));
+
+    errno = xerrno;
+    return -1;
+  }
+
+  return 0;
+}
+
 pr_fs_t *pr_register_fs2(pool *p, const char *name, const char *path,
     int flags) {
   pr_fs_t *fs = NULL;
